### PR TITLE
[FLINK-19404][python] Support Pandas Stream Over Window Aggregation

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change 

*This pull request will support Pandas Stream Over Window Aggregation*

## Brief change log

  - *Add `StreamExecPythonOverAggregateRule` and `StreamExecPythonOverAggregate`*
  - *Add `StreamArrowPythonProcTimeRangeBoundedOverWindowAggregateFunctionOperator`*
  - *Add `StreamArrowPythonProcTimeRowsBoundedOverWindowAggregateFunctionOperator`*
  - *Add `StreamArrowPythonRowTimeRangeBoundedOverWindowAggregateFunctionOperator`*
  - *Add `StreamArrowPythonRowTimeRowsBoundedOverWindowAggregateFunctionOperator`*

## Verifying this change

This change added tests and can be verified as follows:
  
  - *Add plan UT `PythonOverWindowAggregateTest`*
  - *Add operator UT `StreamArrowPythonProcTimeRangeBoundedOverWindowAggregateFunctionOperatorTest`*
  - *Add operator UT `StreamArrowPythonProcTimeRowsBoundedOverWindowAggregateFunctionOperatorTest`*
  - *Add operator UT `StreamArrowPythonRowTimeRangeBoundedOverWindowAggregateFunctionOperatorTest`*
  - *Add operator UT `StreamArrowPythonRowTimeRowsBoundedOverWindowAggregateFunctionOperatorTest`*
  - *Add IT test `test_over_rows_window_aggregate_function` and `test_over_range_window_aggregate_function` in `test_pandas_udaf`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecPythonOverAggregate.scala [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+      None
+    } else {
+      throw new TableException(
+        "OVER windows can only be applied on time attributes.")
+    }
+
+    if (overWindow.lowerBound.isPreceding
+      && overWindow.lowerBound.isUnbounded) {
+      throw new TableException(
+        "OVER PRECEDING windows are not supported yet."
```
Comment (dianfu) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecPythonOverAggregate.scala:133]:
```suggestion
        "UNBOUNDED PRECEDING OVER windows are not supported yet for Python UDAF."
```

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecPythonOverAggregate.scala [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+      )
+    } else if (!overWindow.upperBound.isCurrentRow) {
+      throw new TableException(
+        "OVER FOLLOWING windows are not supported yet."
```
Comment (dianfu) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecPythonOverAggregate.scala:137]:
```suggestion
        "UNBOUNDED FOLLOWING OVER windows are not supported yet for Python UDAF."
```

File: flink-python/pyflink/table/tests/test_pandas_udaf.py [hunk 3]
```diff
@@ -504,6 +504,129 @@ def test_tumbling_group_window_over_count(self):
...
+            '2,3,2013-01-01 03:30:00'
+        ]
+        source_path = tmp_dir + '/test_over_range_window_aggregate_function.csv'
+        with open(source_path, 'w') as fd:
+            for ele in data:
+                fd.write(ele + '\n')
+        max_add_min_udaf = udaf(lambda a: a.max() + a.min(),
+                                result_type=DataTypes.SMALLINT(),
+                                func_type='pandas')
+        self.env.set_stream_time_characteristic(TimeCharacteristic.EventTime)
```
Comment (dianfu) [inline flink-python/pyflink/table/tests/test_pandas_udaf.py:527]: Add test cases for processing time?

**Review by dianfu — COMMENTED** [review]
@HuangXingBo Thanks for the PR.

**Comment by HuangXingBo** [comment]
@dianfu Thanks a lot for the review. I have addressed the comments at the latest commit.

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/stream/AbstractStreamArrowPythonBoundedRowsOperator.java [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+			} else {
+				windowData.pop();
+				windowData.add(inputs.get(i));
+			}
+			for (RowData rowData : windowData) {
+				arrowSerializer.write(getFunctionInput(rowData));
+			}
+			currentBatchCount += windowData.size();
+		}
+		if (currentBatchCount > 0) {
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/stream/AbstractStreamArrowPythonBoundedRowsOperator.java:235]: Extract these code into a separate class and share them between rows and ranges over window?

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/stream/AbstractStreamArrowPythonBoundedRangeOperator.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+						List<RowData> dataList = entry.getValue();
+						if (dataList != null) {
+							for (RowData data : dataList) {
+								arrowSerializer.write(getFunctionInput(data));
+								currentBatchCount++;
+							}
+						}
+					}
+				} else {
+					outdatedTs.add(dataTs);
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/stream/AbstractStreamArrowPythonBoundedRangeOperator.java:156]: could clear the state directly, there is no need to store them into a list.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecPythonOverAggregate.scala [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+      None
+    } else {
+      throw new TableException(
+        "OVER windows can only be applied on time attributes.")
+    }
+
+    if (overWindow.lowerBound.isPreceding
+      && overWindow.lowerBound.isUnbounded) {
+      throw new TableException(
+        "Python UDAF are not supported to be used in UNBOUNDED PRECEDING OVER windows."
```
Comment (dianfu) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecPythonOverAggregate.scala:133]:
```suggestion
        "Python UDAF is not supported to be used in UNBOUNDED PRECEDING OVER windows."
```

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecPythonOverAggregate.scala [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+      )
+    } else if (!overWindow.upperBound.isCurrentRow) {
+      throw new TableException(
+        "Python UDAF are not supported to be used in UNBOUNDED FOLLOWING OVER windows."
```
Comment (dianfu) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecPythonOverAggregate.scala:137]: ditto

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/stream/AbstractStreamArrowPythonBoundedRowsOperator.java [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+	}
+
+	void triggerWindowProcess(List<RowData> inputs, int i, int index) throws Exception {
+		int startIndex;
+		int startPos = 0;
+		if (windowData.isEmpty()) {
+			if (i >= lowerBoundary) {
+				for (int j = (int) (i - lowerBoundary); j <= i; j++) {
+					RowData rowData = inputs.get(j);
+					windowData.add(rowData);
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/stream/AbstractStreamArrowPythonBoundedRowsOperator.java:167]: What about just building the windowData list if windowData is empty? I think we can share the same logic about performing the serialization.

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/stream/AbstractStreamArrowPythonBoundedRangeOperator.java [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+			}
+			cleanupTsState.update(maxCleanupTimestamp);
+		}
+	}
+
+	private void triggerWindowProcess(long upperLimit, List<RowData> inputs) throws Exception {
+		long lowerLimit = upperLimit - lowerBoundary;
+		List<Long> outdatedTs = new LinkedList<>();
+		if (inputs != null) {
+
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/stream/AbstractStreamArrowPythonBoundedRangeOperator.java:142]: unnecessary empty line

**Comment by HuangXingBo** [comment]
@dianfu Thanks a lot for the review. I have addressed the comments at the latest commit.

File: flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/stream/AbstractStreamArrowPythonBoundedRowsOperator.java [hunk 10]
```diff
@@ -158,88 +161,46 @@ void registerProcessingCleanupTimer(long currentTime) throws Exception {
...
-						RowData rowData = previousData.get(j);
-						windowData.add(rowData);
-						arrowSerializer.write(getFunctionInput(rowData));
+					ListIterator<RowData> previousDataIter = previousData.listIterator(length);
+					while (previousDataIter.hasPrevious() && remainingDataCount > 0) {
+						windowData.addFirst(previousDataIter.previous());
+						remainingDataCount--;
 						currentBatchCount++;
 					}
-					// clear outdated data.
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/aggregate/arrow/stream/AbstractStreamArrowPythonBoundedRowsOperator.java]: These logic about clear outdated state are missing in the latest PR.

**Review by dianfu — COMMENTED** [review]
@HuangXingBo LGTM overall. Just one comment.

**Comment by HuangXingBo** [comment]
@dianfu Thanks a lot for the review. I have addressed the comments at the latest commit.

**Review by dianfu — APPROVED** [review]
LGTM. The test failure is not related to this PR. Merging...

**Pull request closed without merging**